### PR TITLE
improve "exhausted attempts" diagnostics

### PR DESCRIPTION
**Pull request by tomerd**
motivation: better diagnostics when depdendcy graph fails to resolve. the "exhausted attempts" error message should normally not be encountered, other than in edge cases where the graph is not resolvable. nevertheless, it should be readable by users

changes:
* conslidate the formatting of the "exhausted attempts" error message
* format the "exhausted attempts" more clearly so it is more readable
* update test

rdar://112134269

### CODE REVIEW

**Comment by tomerd** [comment]
@swift-ci smoke test

**Comment by tomerd** [comment]
@swift-ci smoke test

**Review by neonichu — APPROVED** [review]
Do you have an example of how the messages look like before and after?

**Comment by tomerd** [comment]
the test code has it

**Comment by tomerd** [comment]
@swift-ci test

File: Sources/Workspace/Diagnostics.swift [hunk 1]
```diff
@@ -160,6 +160,22 @@ extension Basics.Diagnostic {
     static func localArtifactNotFound(artifactPath: AbsolutePath, targetName: String) -> Self {
         .error("local binary target '\(targetName)' at '\(artifactPath)' does not contain a binary artifact.")
     }
+
+    static func exhaustedAttempts(missing: [PackageReference]) -> Self {
+        let missing = missing.sorted(by: { $0.identity < $1.identity }).map{
```
Comment (MaxDesiatov) [inline Sources/Workspace/Diagnostics.swift:165]: nit: formatter not applied

**Comment by tomerd** [comment]
@swift-ci test

**Pull request merged**